### ISSUE TITLE
feat: Analytical frequency/hessian/eigenmatrix gradients via JAX eigenvalue sensitivity

### ISSUE THREAD

**Issue by ericchansen** [ericchansen/q2mm#221]
## Summary

Add analytical (autodiff) gradient support for the frequency, hessian_element, and eigenmatrix evaluators when using the JAX backend. Currently only the energy evaluator supports analytical gradients — all others fall back to finite-difference approximation (2N+1 evaluations per gradient step).

## Motivation

For a force field with P parameters and N atoms, the finite-difference gradient of a frequency-based objective requires **2P full Hessian evaluations per optimizer step** — each Hessian itself being an O(N²) autodiff computation. For the Rh-enamide system (52 atoms, ~20 parameters), that's 40 Hessian evaluations just for frequencies. Analytical gradients reduce this to **one Jacobian computation** via `jax.jacrev`, giving exact derivatives with no step-size tuning.

This is the single highest-impact optimization for mixed energy+frequency fitting, which is the most common q2mm workflow.

### Why not differentiate through `eigh` directly?

JAX's `jnp.linalg.eigh` supports reverse-mode AD, but its gradient is **undefined for degenerate eigenvalues** — the eigenvectors become non-unique within degenerate subspaces, making the backward pass unstable. Real molecules routinely have degenerate modes (E symmetry in C₃ᵥ, T symmetry in Tₐ, etc.), plus the 6 near-zero translational/rotational modes.

### Preferred approach: Manual eigenvalue sensitivity formula

The classical eigenvalue perturbation result avoids differentiating through `eigh` entirely:

```
dλ_k/dp = v_k^T · (dA/dp) · v_k
```

where `A` is the mass-weighted Hessian, `λ_k` and `v_k` are its eigenvalues/eigenvectors, and `dA/dp` is the Jacobian of the mass-weighted Hessian w.r.t. force field parameters. This formula is well-defined even for degenerate eigenvalues (the sum of eigenvalue derivatives within a degenerate subspace is always correct).

The full chain for frequencies is:

```
params → E(params, coords)                          # energy function (JAX)
      → d²E/dcoords² = H(params)                    # Hessian via jax.hessian
      → dH/dp via jax.jacrev(jax.hessian(...))       # 3rd-order mixed derivative
      → mass-weight dH/dp                            # linear, trivially differentiable
      → dλ_k/dp = v_k^T · mw(dH/dp) · v_k           # eigenvalue sensitivity
      → d(freq_k)/dp = d(freq)/dλ · dλ_k/dp          # chain rule through sqrt/unit conversion
```

### Architecture

**Engine provides one new capability:**
- `supports_analytical_hessian_gradients()` → bool (base returns False; JaxEngine returns True)
- `hessian_and_param_jacobian(mol, ff)` → `(H, dH/dp)` where H is (3N, 3N) and dH/dp is (3N, 3N, n_params)

This is computed as:
```python
jax.jacrev(jax.hessian(energy_fn, argnums=coords), argnums=params)
```

**Each evaluator chains through its own observable:**

| Evaluator | Formula | Notes |
|-----------|---------|-------|
| **FrequencyEvaluator** | `dλ_k/dp = v_k^T · mw(dH/dp) · v_k`, then chain through `√λ → cm⁻¹` | Eigenvalue sensitivity formula |
| **HessianElementEvaluator** | `d(H[i,j])/dp = dH_dp[i, j, :]` | Direct extraction from Jacobian |
| **EigenmatrixEvaluator** | `d(Q^T H Q)/dp = Q^T · (dH/dp) · Q` | QM eigenvectors Q are constant |

Then each evaluator chains through the residual formula:
```
d(score)/dp = -2 · Σᵢ [wᵢ² · (refᵢ - calcᵢ) · d(calcᵢ)/dp]
```

### Why the Hessian Jacobian is the right abstraction

1. **Computed once, used by three evaluators** — frequency, hessian_element, and eigenmatrix all derive from the Hessian
2. **The 3rd-order derivative is unavoidable** — `dH/dp = d³E/(dx² dp)` regardless of approach
3. **Avoids JAX `eigh` backward pass entirely** — eigenvalue sensitivity uses eigenvectors only in the forward direction
4. **Matches the existing pattern** — `energy_and_param_grad()` provides raw `dE/dp`, and the EnergyEvaluator chains through the residual formula

### Edge Cases

| Edge case | Impact | Mitigation |
|-----------|--------|------------|
| **Degenerate eigenvalues** | Eigenvalue sensitivity formula still works (sum over degenerate subspace is well-defined). Individual eigenvector assignment within the subspace is arbitrary but the total eigenvalue derivative is correct. | None needed — the formula handles it. |
| **Near-zero eigenvalues** (6 trivial modes) | `d(freq)/dλ = factor / (2·√\|λ\|)` diverges as `λ → 0` | Regularize: `max(\|λ\|, ε)` in the denominator. These modes are never used as fitting targets anyway. |
| **JIT compilation cost** | 3rd-order derivative trace may be slow for large molecules on first call | Cache compiled function on `JaxHandle` (same pattern as `_coord_hess_fn`) |
| **Numerical precision** | 3rd-order derivatives accumulate more floating-point error | Validate against finite-difference Jacobian in tests |

### Files to modify

1. **`q2mm/backends/base.py`** — Add `supports_analytical_hessian_gradients()` (returns False) and `hessian_and_param_jacobian()` (raises NotImplementedError)
2. **`q2mm/backends/mm/jax_engine.py`** — Implement both methods. Add `_hess_param_jac_fn` to `JaxHandle` dataclass. Use `jax.jit(jax.jacrev(jax.hessian(energy_fn, argnums=0), argnums=1))`
3. **`q2mm/models/hessian.py`** — Add `frequency_param_jacobian(H, dH_dp, symbols)` that returns `(freqs, d_freq_d_params)` using the eigenvalue sensitivity formula
4. **`q2mm/optimizers/evaluators/frequency.py`** — `supports_analytical_gradient()` checks `engine.supports_analytical_hessian_gradients()`. Implement `gradient()` using `hessian_and_param_jacobian` + `frequency_param_jacobian`
5. **`q2mm/optimizers/evaluators/hessian_element.py`** — Same pattern. `gradient()` extracts `dH[i,j]/dp` directly from Jacobian tensor.
6. **`q2mm/optimizers/evaluators/eigenmatrix.py`** — Same pattern. `gradient()` computes `Q^T (dH/dp) Q` where Q are the cached QM eigenvectors.
7. **Tests** — Validate analytical gradients against finite-difference for all three evaluators. Both pure-NumPy tests (synthetic Hessian inputs for `frequency_param_jacobian`) and JAX integration tests (end-to-end with `@pytest.mark.jax`).

### Testing strategy

The gold standard: compute analytical gradient, compare to central finite-difference gradient. For each evaluator:
1. Create a small test system (e.g., 3-atom molecule with known force field)
2. Compute analytical gradient via the new methods
3. Compute FD gradient by perturbing each parameter ±h and re-evaluating
4. Assert `np.allclose(analytical, fd, rtol=1e-4)`

## Relationship to existing issues

- **#152 (End-to-end differentiable FF fitting)** — This implements Phase 3 of that roadmap (differentiable frequency/eigenmatrix objective). Closing this makes #152 substantially closer to completion.
- **#176 (JAX-ReaxFF pattern)** — The `hessian_and_param_jacobian` method is a prerequisite for the vmapped architecture proposed there.

## Prior art

- [DMFF](https://github.com/deepmodeling/DMFF) — JAX-based differentiable FF platform, uses `jax.hessian` + `jnp.linalg.eigh` for vibrational analysis
- [JAX-ReaxFF](https://github.com/cagrikymk/JAX-ReaxFF) — Gradient-based ReaxFF optimization via JAX autodiff
- Eigenvalue sensitivity formula: standard result in structural dynamics (see e.g., Fox & Kapoor 1968, Nelson 1976)
